### PR TITLE
Add transitional support to PaymentProcessor.pay

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
This addresses some issues identified by @artfulrobot in
https://artfulrobot.uk/blog/migrating-custom-civicrm-donation-form-support-sca-and-new-order-api
and ensures that PaymentProcessor.pay can provide a bridge between processors accepting 'whatever random parameter
they traditionally have received' and processors using getters to retrieve standardised parameters.

For example it means calling functions can just pass the 'approved' 'currency' parameter and not worry about
currencyID.

Ideally processors would use ->getCurrency() to retrieve the currency but at least for the next
six months they are likely to be looking for ['currencyID' so we ensure they get it

Before
----------------------------------------
Need to pass in 'currency_id' AND 'currencyID' to ensure all processors will get it. Ditto contributionRecurID vs conribution_recur_id

After
----------------------------------------
Calling code should just pass in the documented parameter.

Technical Details
----------------------------------------
From 5.20 payment processors should start to transition to using getters to retrieve 'sanctioned' variables. This should help the calling code to switch to better practice while they get there.

Comments
----------------------------------------
Not addressed from the blog

'email'                => $input_params['email'], // Without this it has to guess which to use
'paymentIntentID'      => $input_params['paymentIntentID'],
 payment_processor_id does not accept names - you need to look up the integer ID

The first one because we didn't touch any contact details yet (not for reason just because we didn't get there). Not sure what to make of paymentIntentID'

The issue with " payment_processor_id does not accept names " is the weirdness around live vs test. I think @mattwire did some code that 'pretends' test has a different name to live &  I'm tempted to support that rather than 'name might be one of test or live, use 'is_test' to determine which

### CODE REVIEW

File: api/v3/PaymentProcessor.php [hunk 1]
```diff
@@ -186,6 +201,11 @@ function _civicrm_api3_payment_processor_pay_spec(&$params) {
     'title' => ts('Invoice ID'),
     'type' => CRM_Utils_Type::T_STRING,
   ];
+  $params['currency'] = [
+    'title' => ts('Curency'),
```
Comment (mattwire) [inline api/v3/PaymentProcessor.php]: Typo - "Currency"

File: api/v3/PaymentProcessor.php [hunk 2]
```diff
@@ -186,6 +201,11 @@ function _civicrm_api3_payment_processor_pay_spec(&$params) {
...
+    'type' => CRM_Utils_Type::T_STRING,
+    'api.default' => Civi::settings()->get('defaultCurrency'),
+  ];
```
Comment (mattwire) [inline api/v3/PaymentProcessor.php]: Can we add the description parameter here with an example - as the input format has to be of the three letter kind - is it iso4217? https://www.xe.com/pt/iso4217.php

File: api/v3/PaymentProcessor.php (hunk 1 above)
Comment (eileenmcnaughton) [reply api/v3/PaymentProcessor.php]: oh dear

**Comment by eileenmcnaughton** [comment]
@mattwire I've addressed the 2 things you mention. 

I can look at @artfulrobot's 2 point separately but the questions are

1) email - do we just add 'email' as a property/param. Or is it more rightly billingEmail
2) Stripe specific parameters - should we add a new passthroughParameters or should we do something more discoverable - ie. make it so if you call getfields with the payment_processor_id & action = pay it will allow the processor object to alter the spec - hmm - still pondering if that would actually be helpful. I guess then PaymentProcessor.pay would have to process that spec & the calling function would still pass in a flat array which I guess is the goal

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by mattwire** [comment]
> email - do we just add 'email' as a property/param. Or is it more rightly billingEmail

Another one for discussion on a lab ticket I think - but it has to be the "billing" email that they just entered by default - see my version for how my processors do it currently: https://lab.civicrm.org/extensions/mjwshared/blob/master/CRM/Core/Payment/MJWTrait.php#L41-61

> Stripe specific parameters - should we add a new passthroughParameters or should we do

In theory I could probably handle the two "params" (paymentIntentID and paymentMethodID) as a single "token" parameter. However, it's cleaner for me to implement as two separate input parameters and I can see cases where we *could* have a requirement for multiple input parameters - in fact things like credit card details and bank details are all "multiple input" parameters that are specific to the payment processor. Example:
* UK bank account - pass account number / sortcode
* EU bank account - pass IBAN number
* US bank account - pass routing number, SWIFT code etc.

**Comment by mattwire** [comment]
No issues with prefixing custom parameters - but it would certainly be nice to have a "correct" way of setting/getting them - it's a bit of a nightmare especially when making it work for drupal_webform, frontend/backend forms with/without confirmation pages etc.

**Comment by mattwire** [comment]
> Set them using setCustomProperty like this?

Yes I think so. Still not sure how they would persist through our existing multi-step contribution/event forms but we could try and address that later.

**Comment by eileenmcnaughton** [comment]
@mattwire @artfulrobot I think we should merge my getters & setters revert as step   1 & focus on nailing @artfulrobot's alternative (possibly after the rc is cut now) as step 2 & then finish this & the other one

**Pull request closed without merging**